File: crates/codegraph-core/src/extractors/php.rs
use super::helpers::*;
use super::SymbolExtractor;
use crate::cfg::build_function_cfg;
use crate::complexity::compute_all_metrics;
use crate::types::*;
use tree_sitter::{Node, Tree};

pub struct PhpExtractor;

impl SymbolExtractor for PhpExtractor {
    fn extract(&self, tree: &Tree, source: &[u8], file_path: &str) -> FileSymbols {
        let mut symbols = FileSymbols::new(file_path.to_string());
        walk_tree(&tree.root_node(), source, &mut symbols, match_php_node);
        walk_tree(&tree.root_node(), source, &mut symbols, match_php_type_map);
        walk_ast_nodes_with_config(&tree.root_node(), source, &mut symbols.ast_nodes, &PHP_AST_CONFIG);
        symbols
    }
}

const PHP_CLASS_KINDS: &[&str] = &["class_declaration", "trait_declaration", "enum_declaration"];

fn find_php_parent_class(node: &Node, source: &[u8]) -> Option<String> {
    find_enclosing_type_name(node, PHP_CLASS_KINDS, source)
}

fn match_php_node(node: &Node, source: &[u8], symbols: &mut FileSymbols, _depth: usize) {
    match node.kind() {
        "function_definition" => handle_function_def(node, source, symbols),
        "class_declaration" => handle_class_decl(node, source, symbols),
        "interface_declaration" => handle_interface_decl(node, source, symbols),
        "trait_declaration" => handle_trait_decl(node, source, symbols),
        "enum_declaration" => handle_enum_decl(node, source, symbols),
        "method_declaration" => handle_method_decl(node, source, symbols),
        "namespace_use_declaration" => handle_namespace_use(node, source, symbols),
        "function_call_expression" => handle_function_call(node, source, symbols),
        "member_call_expression" => handle_member_call(node, source, symbols),
        "scoped_call_expression" => handle_scoped_call(node, source, symbols),
        "object_creation_expression" => handle_object_creation(node, source, symbols),
        _ => {}
    }
}

// ── Per-node-kind handlers for walk_node_depth ───────────────────────────────

fn handle_function_def(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    if let Some(name_node) = node.child_by_field_name("name") {
        let children = extract_php_parameters(node, source);
        symbols.definitions.push(Definition {
            name: node_text(&name_node, source).to_string(),
            kind: "function".to_string(),
            line: start_line(node),
            end_line: Some(end_line(node)),
            decorators: None,
            complexity: compute_all_metrics(node, source, "php"),
            cfg: build_function_cfg(node, "php", source),
            children: opt_children(children),
        });
    }
}

fn handle_class_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let Some(name_node) = node.child_by_field_name("name") else { return };
    let class_name = node_text(&name_node, source).to_string();
    let children = extract_php_class_properties(node, source);
    symbols.definitions.push(Definition {
        name: class_name.clone(),
        kind: "class".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: opt_children(children),
    });

    // Extends
    let base_clause = node
        .child_by_field_name("base_clause")
        .or_else(|| find_child(node, "base_clause"));
    if let Some(base_clause) = base_clause {
        for i in 0..base_clause.child_count() {
            if let Some(child) = base_clause.child(i) {
                if child.kind() == "name" || child.kind() == "qualified_name" {
                    symbols.classes.push(ClassRelation {
                        name: class_name.clone(),
                        extends: Some(node_text(&child, source).to_string()),
                        implements: None,
                        line: start_line(node),
                    });
                    break;
                }
            }
        }
    }

    // Implements
    let interface_clause = find_child(node, "class_interface_clause");
    if let Some(interface_clause) = interface_clause {
        for i in 0..interface_clause.child_count() {
            if let Some(child) = interface_clause.child(i) {
                if child.kind() == "name" || child.kind() == "qualified_name" {
                    symbols.classes.push(ClassRelation {
                        name: class_name.clone(),
                        extends: None,
                        implements: Some(node_text(&child, source).to_string()),
                        line: start_line(node),
                    });
                }
            }
        }
    }
}

fn handle_interface_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let Some(name_node) = node.child_by_field_name("name") else { return };
    let iface_name = node_text(&name_node, source).to_string();
    symbols.definitions.push(Definition {
        name: iface_name.clone(),
        kind: "interface".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: None,
    });
    if let Some(body) = node.child_by_field_name("body") {
        for i in 0..body.child_count() {
            let Some(child) = body.child(i) else { continue };
            if child.kind() != "method_declaration" { continue; }
            if let Some(meth_name) = child.child_by_field_name("name") {
                symbols.definitions.push(Definition {
                    name: format!("{}.{}", iface_name, node_text(&meth_name, source)),
                    kind: "method".to_string(),
                    line: start_line(&child),
                    end_line: Some(end_line(&child)),
                    decorators: None,
                    complexity: compute_all_metrics(&child, source, "php"),
                    cfg: build_function_cfg(&child, "php", source),
                    children: None,
                });
            }
        }
    }
}

fn handle_trait_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    if let Some(name_node) = node.child_by_field_name("name") {
        symbols.definitions.push(Definition {
            name: node_text(&name_node, source).to_string(),
            kind: "trait".to_string(),
            line: start_line(node),
            end_line: Some(end_line(node)),
            decorators: None,
            complexity: None,
            cfg: None,
            children: None,
        });
    }
}

fn handle_enum_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    if let Some(name_node) = node.child_by_field_name("name") {
        let enum_name = node_text(&name_node, source).to_string();
        let children = extract_php_enum_cases(node, source);
        symbols.definitions.push(Definition {
            name: enum_name,
            kind: "enum".to_string(),
            line: start_line(node),
            end_line: Some(end_line(node)),
            decorators: None,
            complexity: None,
            cfg: None,
            children: opt_children(children),
        });
    }
}

fn handle_method_decl(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    if let Some(name_node) = node.child_by_field_name("name") {
        let parent_class = find_php_parent_class(node, source);
        let name = node_text(&name_node, source);
        let full_name = match &parent_class {
            Some(cls) => format!("{}.{}", cls, name),
            None => name.to_string(),
        };
        let children = extract_php_parameters(node, source);
        symbols.definitions.push(Definition {
            name: full_name,
            kind: "method".to_string(),
            line: start_line(node),
            end_line: Some(end_line(node)),
            decorators: None,
            complexity: compute_all_metrics(node, source, "php"),
            cfg: build_function_cfg(node, "php", source),
            children: opt_children(children),
        });
    }
}

fn handle_namespace_use(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    for i in 0..node.child_count() {
        let Some(child) = node.child(i) else { continue };
        if child.kind() == "namespace_use_clause" {
            let name_node = find_child(&child, "qualified_name")
                .or_else(|| find_child(&child, "name"));
            if let Some(name_node) = name_node {
                let full_path = node_text(&name_node, source).to_string();
                let last_name = full_path.split('\\').last().unwrap_or("").to_string();
                let alias = child.child_by_field_name("alias");
                let alias_text = alias
                    .map(|a| node_text(&a, source).to_string())
                    .unwrap_or(last_name);
                let mut imp = Import::new(full_path, vec![alias_text], start_line(node));
                imp.php_use = Some(true);
                symbols.imports.push(imp);
            }
        }
        // Single use clause without wrapper
        if child.kind() == "qualified_name" || child.kind() == "name" {
            let full_path = node_text(&child, source).to_string();
            let last_name = full_path.split('\\').last().unwrap_or("").to_string();
            let mut imp = Import::new(full_path, vec![last_name], start_line(node));
            imp.php_use = Some(true);
            symbols.imports.push(imp);
        }
    }
}

fn handle_function_call(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let fn_node = node.child_by_field_name("function").or_else(|| node.child(0));
    let Some(fn_node) = fn_node else { return };
    match fn_node.kind() {
        "name" | "identifier" => {
            symbols.calls.push(Call {
                name: node_text(&fn_node, source).to_string(),
                line: start_line(node),
                dynamic: None,
                receiver: None,
            });
        }
        "qualified_name" => {
            let text = node_text(&fn_node, source);
            let last = text.split('\\').last().unwrap_or("");
            symbols.calls.push(Call {
                name: last.to_string(),
                line: start_line(node),
                dynamic: None,
                receiver: None,
            });
        }
        _ => {}
    }
}

fn handle_member_call(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    if let Some(name) = node.child_by_field_name("name") {
        let receiver = node.child_by_field_name("object")
            .map(|obj| node_text(&obj, source).to_string());
        symbols.calls.push(Call {
            name: node_text(&name, source).to_string(),
            line: start_line(node),
            dynamic: None,
            receiver,
        });
    }
}

fn handle_scoped_call(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    if let Some(name) = node.child_by_field_name("name") {
        let receiver = node.child_by_field_name("scope")
            .map(|s| node_text(&s, source).to_string());
        symbols.calls.push(Call {
            name: node_text(&name, source).to_string(),
            line: start_line(node),
            dynamic: None,
            receiver,
        });
    }
}

fn handle_object_creation(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let Some(class_node) = node.child(1) else { return };
    if class_node.kind() != "name" && class_node.kind() != "qualified_name" { return; }
    let text = node_text(&class_node, source);
    let last = text.split('\\').last().unwrap_or("");
    symbols.calls.push(Call {
        name: last.to_string(),
        line: start_line(node),
        dynamic: None,
        receiver: None,
    });
}

// ── Extended kinds helpers ──────────────────────────────────────────────────

fn extract_php_parameters(node: &Node, source: &[u8]) -> Vec<Definition> {
    let mut params = Vec::new();
    let params_node = node.child_by_field_name("parameters")
        .or_else(|| find_child(node, "formal_parameters"));
    if let Some(params_node) = params_node {
        for i in 0..params_node.child_count() {
            if let Some(child) = params_node.child(i) {
                if child.kind() == "simple_parameter"
                    || child.kind() == "variadic_parameter"
                    || child.kind() == "property_promotion_parameter"
                {
                    if let Some(name_node) = child.child_by_field_name("name") {
                        params.push(child_def(
                            node_text(&name_node, source).to_string(),
                            "parameter",
                            start_line(&child),
                        ));
                    }
                }
            }
        }
    }
    params
}

fn extract_php_class_properties(node: &Node, source: &[u8]) -> Vec<Definition> {
    let mut props = Vec::new();
    let body = node.child_by_field_name("body")
        .or_else(|| find_child(node, "declaration_list"));
    if let Some(body) = body {
        for i in 0..body.child_count() {
            if let Some(child) = body.child(i) {
                if child.kind() == "property_declaration" {
                    // Walk property_element children
                    for j in 0..child.child_count() {
                        if let Some(elem) = child.child(j) {
                            if elem.kind() == "property_element" {
                                if let Some(name_node) = elem.child(0) {
                                    if name_node.kind() == "variable_name" {
                                        props.push(child_def(
                                            node_text(&name_node, source).to_string(),
                                            "property",
                                            start_line(&child),
                                        ));
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    props
}

fn extract_php_enum_cases(node: &Node, source: &[u8]) -> Vec<Definition> {
    let mut cases = Vec::new();
    let body = node.child_by_field_name("body")
        .or_else(|| find_child(node, "enum_declaration_list"));
    if let Some(body) = body {
        for i in 0..body.child_count() {
            if let Some(child) = body.child(i) {
                if child.kind() == "enum_case" {
                    if let Some(name_node) = child.child_by_field_name("name") {
                        cases.push(child_def(
                            node_text(&name_node, source).to_string(),
                            "constant",
                            start_line(&child),
                        ));
                    }
                }
            }
        }
    }
    cases
}

fn extract_php_type_name<'a>(type_node: &Node<'a>, source: &'a [u8]) -> Option<&'a str> {
    match type_node.kind() {
        "named_type" | "name" | "qualified_name" => Some(node_text(type_node, source)),
        "optional_type" => {
            // ?MyType → skip the ? and get inner type
            type_node.child(1)
                .or_else(|| type_node.child(0))
                .and_then(|inner| extract_php_type_name(&inner, source))
        }
        // Skip union/intersection types (too ambiguous)
        "union_type" | "intersection_type" => None,
        _ => None,
    }
}

fn match_php_type_map(node: &Node, source: &[u8], symbols: &mut FileSymbols, _depth: usize) {
    match node.kind() {
        "simple_parameter" | "variadic_parameter" | "property_promotion_parameter" => {
            if let Some(type_node) = node.child_by_field_name("type") {
                if let Some(type_name) = extract_php_type_name(&type_node, source) {
                    let name_node = node.child_by_field_name("name")
                        .or_else(|| find_child(node, "variable_name"));
                    if let Some(name_node) = name_node {
                        symbols.type_map.push(TypeMapEntry {
                            name: node_text(&name_node, source).to_string(),
                            type_name: type_name.to_string(),
                        });
                    }
                }
            }
        }
        _ => {}
    }
}

File: src/extractors/php.ts
import type {
  Call,
  ExtractorOutput,
  SubDeclaration,
  TreeSitterNode,
  TreeSitterTree,
} from '../types.js';
import {
  extractBodyMembers,
  extractModifierVisibility,
  findChild,
  lastPathSegment,
  MAX_WALK_DEPTH,
  nodeEndLine,
} from './helpers.js';

function extractPhpParameters(fnNode: TreeSitterNode): SubDeclaration[] {
  const params: SubDeclaration[] = [];
  const paramsNode =
    fnNode.childForFieldName('parameters') || findChild(fnNode, 'formal_parameters');
  if (!paramsNode) return params;
  for (let i = 0; i < paramsNode.childCount; i++) {
    const param = paramsNode.child(i);
    if (!param) continue;
    if (param.type === 'simple_parameter' || param.type === 'variadic_parameter') {
      const nameNode = param.childForFieldName('name') || findChild(param, 'variable_name');
      if (nameNode) {
        params.push({ name: nameNode.text, kind: 'parameter', line: param.startPosition.row + 1 });
      }
    }
  }
  return params;
}

/** Extract property declarations from a PHP class member. */
function extractPhpProperties(member: TreeSitterNode, children: SubDeclaration[]): void {
  for (let j = 0; j < member.childCount; j++) {
    const el = member.child(j);
    if (!el || el.type !== 'property_element') continue;
    const varNode = findChild(el, 'variable_name');
    if (varNode) {
      children.push({
        name: varNode.text,
        kind: 'property',
        line: member.startPosition.row + 1,
        visibility: extractModifierVisibility(member),
      });
    }
  }
}

/** Extract constant declarations from a PHP class member. */
function extractPhpConstants(member: TreeSitterNode, children: SubDeclaration[]): void {
  for (let j = 0; j < member.childCount; j++) {
    const el = member.child(j);
    if (!el || el.type !== 'const_element') continue;
    const nameNode = el.childForFieldName('name') || findChild(el, 'name');
    if (nameNode) {
      children.push({
        name: nameNode.text,
        kind: 'constant',
        line: member.startPosition.row + 1,
      });
    }
  }
}

function extractPhpClassChildren(classNode: TreeSitterNode): SubDeclaration[] {
  const children: SubDeclaration[] = [];
  const body = classNode.childForFieldName('body') || findChild(classNode, 'declaration_list');
  if (!body) return children;
  for (let i = 0; i < body.childCount; i++) {
    const member = body.child(i);
    if (!member) continue;
    if (member.type === 'property_declaration') {
      extractPhpProperties(member, children);
    } else if (member.type === 'const_declaration') {
      extractPhpConstants(member, children);
    }
  }
  return children;
}

function extractPhpEnumCases(enumNode: TreeSitterNode): SubDeclaration[] {
  return extractBodyMembers(enumNode, ['body', 'enum_declaration_list'], 'enum_case', 'constant');
}

/**
 * Extract symbols from PHP files.
 */
export function extractPHPSymbols(tree: TreeSitterTree, _filePath: string): ExtractorOutput {
  const ctx: ExtractorOutput = {
    definitions: [],
    calls: [],
    imports: [],
    classes: [],
    exports: [],
    typeMap: new Map(),
  };

  walkPhpNode(tree.rootNode, ctx);
  extractPhpTypeMap(tree.rootNode, ctx);
  return ctx;
}

function walkPhpNode(node: TreeSitterNode, ctx: ExtractorOutput): void {
  switch (node.type) {
    case 'function_definition':
      handlePhpFuncDef(node, ctx);
      break;
    case 'class_declaration':
      handlePhpClassDecl(node, ctx);
      break;
    case 'interface_declaration':
      handlePhpInterfaceDecl(node, ctx);
      break;
    case 'trait_declaration':
      handlePhpTraitDecl(node, ctx);
      break;
    case 'enum_declaration':
      handlePhpEnumDecl(node, ctx);
      break;
    case 'method_declaration':
      handlePhpMethodDecl(node, ctx);
      break;
    case 'namespace_use_declaration':
      handlePhpNamespaceUse(node, ctx);
      break;
    case 'function_call_expression':
      handlePhpFuncCall(node, ctx);
      break;
    case 'member_call_expression':
      handlePhpMemberCall(node, ctx);
      break;
    case 'scoped_call_expression':
      handlePhpScopedCall(node, ctx);
      break;
    case 'object_creation_expression':
      handlePhpObjectCreation(node, ctx);
      break;
  }

  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (child) walkPhpNode(child, ctx);
  }
}

// ── Walk-path per-node-type handlers ────────────────────────────────────────

function handlePhpFuncDef(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  const params = extractPhpParameters(node);
  ctx.definitions.push({
    name: nameNode.text,
    kind: 'function',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
    children: params.length > 0 ? params : undefined,
  });
}

function handlePhpClassDecl(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  const classChildren = extractPhpClassChildren(node);
  ctx.definitions.push({
    name: nameNode.text,
    kind: 'class',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
    children: classChildren.length > 0 ? classChildren : undefined,
  });
  const baseClause = node.childForFieldName('base_clause') || findChild(node, 'base_clause');
  if (baseClause) {
    for (let i = 0; i < baseClause.childCount; i++) {
      const child = baseClause.child(i);
      if (child && (child.type === 'name' || child.type === 'qualified_name')) {
        ctx.classes.push({
          name: nameNode.text,
          extends: child.text,
          line: node.startPosition.row + 1,
        });
        break;
      }
    }
  }
  const interfaceClause = findChild(node, 'class_interface_clause');
  if (interfaceClause) {
    for (let i = 0; i < interfaceClause.childCount; i++) {
      const child = interfaceClause.child(i);
      if (child && (child.type === 'name' || child.type === 'qualified_name')) {
        ctx.classes.push({
          name: nameNode.text,
          implements: child.text,
          line: node.startPosition.row + 1,
        });
      }
    }
  }
}

function handlePhpInterfaceDecl(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  ctx.definitions.push({
    name: nameNode.text,
    kind: 'interface',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
  });
  const body = node.childForFieldName('body');
  if (body) {
    for (let i = 0; i < body.childCount; i++) {
      const child = body.child(i);
      if (child && child.type === 'method_declaration') {
        const methName = child.childForFieldName('name');
        if (methName) {
          ctx.definitions.push({
            name: `${nameNode.text}.${methName.text}`,
            kind: 'method',
            line: child.startPosition.row + 1,
            endLine: child.endPosition.row + 1,
          });
        }
      }
    }
  }
}

function handlePhpTraitDecl(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  ctx.definitions.push({
    name: nameNode.text,
    kind: 'trait',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
  });
}

function handlePhpEnumDecl(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  const enumChildren = extractPhpEnumCases(node);
  ctx.definitions.push({
    name: nameNode.text,
    kind: 'enum',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
    children: enumChildren.length > 0 ? enumChildren : undefined,
  });
}

function handlePhpMethodDecl(node: TreeSitterNode, ctx: ExtractorOutput): void {
  // Skip interface methods already emitted by handlePhpInterfaceDecl
  if (node.parent?.parent?.type === 'interface_declaration') return;
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  const parentClass = findPHPParentClass(node);
  const fullName = parentClass ? `${parentClass}.${nameNode.text}` : nameNode.text;
  const params = extractPhpParameters(node);
  ctx.definitions.push({
    name: fullName,
    kind: 'method',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
    children: params.length > 0 ? params : undefined,
    visibility: extractModifierVisibility(node),
  });
}

function handlePhpNamespaceUse(node: TreeSitterNode, ctx: ExtractorOutput): void {
  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (child && child.type === 'namespace_use_clause') {
      const nameNode = findChild(child, 'qualified_name') || findChild(child, 'name');
      if (nameNode) {
        const fullPath = nameNode.text;
        const lastName = lastPathSegment(fullPath, '\\');
        const alias = child.childForFieldName('alias');
        ctx.imports.push({
          source: fullPath,
          names: [alias ? alias.text : lastName],
          line: node.startPosition.row + 1,
          phpUse: true,
        });
      }
    }
    if (child && (child.type === 'qualified_name' || child.type === 'name')) {
      const fullPath = child.text;
      const lastName = lastPathSegment(fullPath, '\\');
      ctx.imports.push({
        source: fullPath,
        names: [lastName],
        line: node.startPosition.row + 1,
        phpUse: true,
      });
    }
  }
}

function handlePhpFuncCall(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const fn = node.childForFieldName('function') || node.child(0);
  if (!fn) return;
  if (fn.type === 'name' || fn.type === 'identifier') {
    ctx.calls.push({ name: fn.text, line: node.startPosition.row + 1 });
  } else if (fn.type === 'qualified_name') {
    const parts = fn.text.split('\\');
    ctx.calls.push({ name: parts[parts.length - 1] ?? fn.text, line: node.startPosition.row + 1 });
  }
}

function handlePhpMemberCall(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const name = node.childForFieldName('name');
  if (!name) return;
  const obj = node.childForFieldName('object');
  const call: Call = { name: name.text, line: node.startPosition.row + 1 };
  if (obj) call.receiver = obj.text;
  ctx.calls.push(call);
}

function handlePhpScopedCall(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const name = node.childForFieldName('name');
  if (!name) return;
  const scope = node.childForFieldName('scope');
  const call: Call = { name: name.text, line: node.startPosition.row + 1 };
  if (scope) call.receiver = scope.text;
  ctx.calls.push(call);
}

function handlePhpObjectCreation(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const classNode = node.child(1);
  if (classNode && (classNode.type === 'name' || classNode.type === 'qualified_name')) {
    const parts = classNode.text.split('\\');
    ctx.calls.push({
      name: parts[parts.length - 1] ?? classNode.text,
      line: node.startPosition.row + 1,
    });
  }
}

function extractPhpTypeMap(node: TreeSitterNode, ctx: ExtractorOutput): void {
  extractPhpTypeMapDepth(node, ctx, 0);
}

function extractPhpTypeMapDepth(node: TreeSitterNode, ctx: ExtractorOutput, depth: number): void {
  if (depth >= MAX_WALK_DEPTH) return;

  // Function/method parameters with type hints
  if (
    node.type === 'simple_parameter' ||
    node.type === 'variadic_parameter' ||
    node.type === 'property_promotion_parameter'
  ) {
    const typeNode = node.childForFieldName('type');
    const nameNode = node.childForFieldName('name') || findChild(node, 'variable_name');
    if (typeNode && nameNode) {
      const typeName = extractPhpTypeName(typeNode);
      if (typeName) ctx.typeMap?.set(nameNode.text, { type: typeName, confidence: 0.9 });
    }
  }

  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (child) extractPhpTypeMapDepth(child, ctx, depth + 1);
  }
}

function extractPhpTypeName(typeNode: TreeSitterNode): string | null {
  if (!typeNode) return null;
  const t = typeNode.type;
  if (t === 'named_type' || t === 'name' || t === 'qualified_name') return typeNode.text;
  // Nullable: ?MyType
  if (t === 'optional_type') {
    const inner = typeNode.child(1) || typeNode.child(0);
    return inner ? extractPhpTypeName(inner) : null;
  }
  // Skip union types (too ambiguous)
  if (t === 'union_type' || t === 'intersection_type') return null;
  return null;
}

function findPHPParentClass(node: TreeSitterNode): string | null {
  let current = node.parent;
  while (current) {
    if (
      current.type === 'class_declaration' ||
      current.type === 'trait_declaration' ||
      current.type === 'enum_declaration'
    ) {
      const nameNode = current.childForFieldName('name');
      return nameNode ? nameNode.text : null;
    }
    current = current.parent;
  }
  return null;
}
